Software Requirements
=====================
This course requires a variety of software packages. To ensure students have
as similar results to others, this course will use specific versions of those
software. The software, version and their purpose are provided below.

.. list-table::
   :widths: 25 25 50
   :header-rows: 1

   * - Software
     - Version
     - Purpose
   * - `R <https://www.r-project.org/>`__
     - 4.1.3
     - A free software environment for statistical computing and graphics.
   * - `FastQC <https://www.bioinformatics.babraham.ac.uk/projects/fastqc/>`__
     - 0.11.9
     - A quality control tool for high throughput sequence data.
   * - `MultiQC <https://multiqc.info/>`__
     - 1.13a
     - MultiQC summarizes the output from numerous bioinformatics tools into a single report.
   * - `SRA Tools <https://github.com/ncbi/sra-tools>`__
     - 2.11.0
     - A collection of tools and libraries for using data in the INSDC Sequence Read Archives.
   * - `FastP <https://github.com/OpenGene/fastp>`__
     - 0.23.2
     - Provides fast all-in-one preprocessing for FastQ file.
   * - `SamTools <http://www.htslib.org/>`__
     - 1.15.1
     - A suite of programs for interacting with high-throughput sequencing data.
   * - `NOVOPlasty <https://github.com/ndierckx/NOVOPlasty/>`__
     - 4.3.1
     - A de novo assembler and heteroplasmy/variance caller for short circular genomes.
   * - `BWA <http://bio-bwa.sourceforge.net/>`__
     - 0.7.17
     - A software package for mapping low-divergent sequences against a large reference genome.
   * - `Jellyfish <https://github.com/gmarcais/Jellyfish>`__
     - 2.2.10
     - A tool for fast, memory-efficient counting of k-mers in DNA.
   * - `Hifiasm <https://github.com/chhylp123/hifiasm>`__
     - 0.16.1
     - Hifiasm is a fast haplotype-resolved de novo assembler for PacBio HiFi reads.
   * - `Mummer <http://mummer.sourceforge.net/>`__
     - 3.23
     - Ultra-fast alignment of large-scale DNA and protein sequences.
   * - `BedTools <https://bedtools.readthedocs.io/en/latest/>`__
     - 2.30.0
     - A swiss-army knife of tools for a wide-range of genomics analysis tasks.
   * - `RepeatMasker <https://www.repeatmasker.org/>`__
     - 4.1.2.p1
     - Screens DNA sequences for interspersed repeats and low complexity DNA sequences
   * - `Braker2 <https://github.com/Gaius-Augustus/BRAKER>`__
     - 2.1.6
     - Gene structural annotation.
   * - `BUSCO <https://busco.ezlab.org/>`__
     - 4.1.2
     - Assesses genome assembly and annotation completeness using single-copy orthologs.
   * - `EDTA <https://github.com/oushujun/EDTA>`__
     - 2.0.1
     - Automated whole-genome de-novo TE annotation and benchmarking.
   * - `hic_qc <https://github.com/phasegenomics/hic_qc>`__
     - git commit 6881c33
     - Performs QC Checks for Hi-C libraries using reads in a BAM file aligned to the genome assembly. The
       version of this software is a commit to the source repository made on June 27, 2022.
